internal/topology/variables/clusterclass_variable_validation.go

```diff
@@ -215,9 +225,128 @@ func validateSchema(schema *apiextensions.JSONSchemaProps, fldPath *field.Path)
+	}
+	if typeInfo == nil {
+		return allErrs
 	}
 
+	compResults, err := cel.Compile(
+		typeInfo.Schema,
+		typeInfo.DeclType,
+		celconfig.PerCallLimit,
+		environment.MustBaseEnvSet(environment.DefaultCompatibilityVersion()),
+		cel.NewExpressionsEnvLoader(),
+	)
+	if err != nil {
+		return append(allErrs, field.InternalError(fldPath.Child("x-kubernetes-validations"), err))
+	}
+
+	for i, cr := range compResults {
+		expressionCost := getExpressionCost(cr, celContext)
+		if expressionCost > StaticEstimatedCostLimit {
+			costErrorMsg := getCostErrorMessage("estimated rule cost", expressionCost, StaticEstimatedCostLimit)
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("x-kubernetes-validations").Index(i).Child("rule"), costErrorMsg))
+		}
+		if celContext.TotalCost != nil {
+			celContext.TotalCost.ObserveExpressionCost(fldPath.Child("x-kubernetes-validations").Index(i).Child("rule"), expressionCost)
+		}
+		if cr.Error != nil {
+			if cr.Error.Type == apiservercel.ErrorTypeRequired {
+				allErrs = append(allErrs, field.Required(fldPath.Child("x-kubernetes-validations").Index(i).Child("rule"), cr.Error.Detail))
+			} else {
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("x-kubernetes-validations").Index(i).Child("rule"), schema.XValidations[i], cr.Error.Detail))
+			}
+		}
+		if cr.MessageExpressionError != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("x-kubernetes-validations").Index(i).Child("messageExpression"), schema.XValidations[i], cr.MessageExpressionError.Detail))
+		} else if cr.MessageExpression != nil {
+			if cr.MessageExpressionMaxCost > StaticEstimatedCostLimit {
+				costErrorMsg := getCostErrorMessage("estimated messageExpression cost", cr.MessageExpressionMaxCost, StaticEstimatedCostLimit)
+				allErrs = append(allErrs, field.Forbidden(fldPath.Child("x-kubernetes-validations").Index(i).Child("messageExpression"), costErrorMsg))
+			}
+			if celContext.TotalCost != nil {
+				celContext.TotalCost.ObserveExpressionCost(fldPath.Child("x-kubernetes-validations").Index(i).Child("messageExpression"), cr.MessageExpressionMaxCost)
+			}
+		}
+	}
 	return allErrs
 }
+
+// multiplyWithOverflowGuard returns the product of baseCost and cardinality unless that product
+// would exceed math.MaxUint, in which case math.MaxUint is returned.
+func multiplyWithOverflowGuard(baseCost, cardinality uint64) uint64 {
+	if baseCost == 0 {
+		// an empty rule can return 0, so guard for that here
+		return 0
+	} else if math.MaxUint/baseCost < cardinality {
+		return math.MaxUint
+	}
+	return baseCost * cardinality
+}
+
+// unbounded uses nil to represent an unbounded cardinality value.
+var unbounded *uint64 = nil //nolint:revive // Using as a named variable to provide the meaning of nil in this context.
+
+func getExpressionCost(cr cel.CompilationResult, cardinalityCost *apiextensionsvalidation.CELSchemaContext) uint64 {
```

Comment:
Let's add some godoc with an explanation how we're calculating the expression cost (not sure what cr.MaxCost and cartinalityCost are exactly)